applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp

```diff
@@ -0,0 +1,336 @@
+
+// ------------------------------------------------------------------------------
+// Project includes
+// ------------------------------------------------------------------------------
+#include "adjoint_local_stress_response_function.h"
+
+namespace Kratos
+{
+    /// Default constructor.
+    AdjointLocalStressResponseFunction::AdjointLocalStressResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
+    {
+        ModelPart& r_model_part = this->GetModelPart();
+
+        ResponseData stress_response_data;
+
+        // Get traced element
+        mIdOfTracedElement = ResponseSettings["traced_element_id"].GetInt();
+        mpTracedElement = r_model_part.pGetElement(mIdOfTracedElement);
+
+        // Tell traced element the stress type
+        TracedStressType traced_stress_type = stress_response_data.ConvertStressType(ResponseSettings["stress_type"].GetString()); 
+        KRATOS_ERROR_IF(traced_stress_type == TracedStressType::StressTypeNotAvailable) << "Chosen stress type is not available!" << std::endl;
+        mpTracedElement->SetValue(TRACED_STRESS_TYPE, static_cast<int>(traced_stress_type) );		
+
+        // Get info how and where to treat the stress
+        mStressTreatment = stress_response_data.ConvertStressTreatment( ResponseSettings["stress_treatment"].GetString() );
+        KRATOS_ERROR_IF(mStressTreatment == StressTreatment::StressTreatmentNotAvailable) << "Chosen option for stress treatmeant is not available! Chose 'GP','node' or 'mean'!" << std::endl;
+
+        if(mStressTreatment == StressTreatment::GaussPoint || mStressTreatment == StressTreatment::Node)
+        {
+            mIdOfLocation = ResponseSettings["stress_location"].GetInt();
+            KRATOS_ERROR_IF(mIdOfLocation < 1) << "Chose a 'stress_location' > 0. Specified 'stress_location': " << mIdOfLocation << std::endl;
+        }
+
+        mStressValue = 0.0;
+    }
+
+    /// Destructor.
+    AdjointLocalStressResponseFunction::~AdjointLocalStressResponseFunction(){}
+
+    // ==============================================================================
+    double AdjointLocalStressResponseFunction::CalculateValue(ModelPart& rModelPart)
+    {
+        KRATOS_TRY;
+
+        // Working variables
+        ProcessInfo &r_current_precess_info = rModelPart.GetProcessInfo();
+        Vector element_stress;
+
+        if(mStressTreatment == StressTreatment::Mean || mStressTreatment == StressTreatment:: GaussPoint)
+            mpTracedElement->Calculate(STRESS_ON_GP, element_stress, r_current_precess_info);
+        else
+            mpTracedElement->Calculate(STRESS_ON_NODE, element_stress, r_current_precess_info);
+
+        const SizeType stress_vec_size = element_stress.size();
+
+        if(mStressTreatment == StressTreatment::Mean)
+        {
+            for(IndexType i = 0; i < stress_vec_size; ++i)
+                mStressValue += element_stress[i];
```

Comment:
You have a serious bug here and it wasn't easy to spot because it's hidden in a long and complicated function that performs multiple operations which are interlaced with each other. This is one of many reasons to refactor this and the three other functions just like this one in this class into smaller ones.

Reply:
Are you talking about the `CalculateValue` function? I can't spot a bug tbh.
Also this function has 40 lines, i wouldn't consider that long.
However i agree that there is room for some refactoring in this class.

Reply:
Discussed in person / done